File: sdks/dotnet/tests/Croniq.Runner.Sdk.Conformance.Tests/HandlerSentinels.cs
using Croniq.Runner.Sdk;
using Croniq.Runner.Sdk.DependencyInjection;
using Croniq.Runner.Sdk.Logging;

using Microsoft.Extensions.Logging;

namespace Croniq.Runner.Sdk.Conformance.Tests;

/// <summary>
/// Translates a YAML handler <c>behavior</c> into a Croniq handler delegate.
/// Every conformance case binding must implement the same sentinels:
/// <c>noop</c>, <c>throw</c>, <c>sleep</c>, <c>log</c>, <c>stream_logs</c>.
/// </summary>
internal static class HandlerSentinels
{
    public static ICroniqRunnerBuilder ApplyTo(ICroniqRunnerBuilder builder, IList<HandlerSpec> handlers)
    {
        foreach (var spec in handlers)
        {
            var snapshot = spec; // capture
            Func<CroniqExecutionContext, CancellationToken, Task> handler = snapshot.Behavior switch
            {
                "noop" => static (_, _) => Task.CompletedTask,
                "throw" => (_, _) => throw new InvalidOperationException(snapshot.ErrorMessage ?? "thrown by conformance handler"),
                "sleep" => async (_, ct) => await Task.Delay(snapshot.DurationMs ?? 0, ct).ConfigureAwait(false),
                "log" => (ctx, _) =>
                {
                    var level = ParseLevel(snapshot.Level);
                    var count = snapshot.Count ?? 1;
                    for (var i = 0; i < count; i++)
                    {
#pragma warning disable CA2254 // template not constant — conformance shim only
                        ctx.Logger.Log(level, snapshot.Message ?? "");
#pragma warning restore CA2254
                    }
                    return Task.CompletedTask;
                }
                ,
                "stream_logs" => async (ctx, ct) =>
                {
                    var count = snapshot.Count ?? 1;
                    var interval = snapshot.IntervalMs ?? 0;
                    var level = ParseLevel(snapshot.Level);
                    var writer = ctx.LogWriter;
                    for (var i = 0; i < count; i++)
                    {
                        await writer.WriteAsync(level, $"line {i + 1}", cancellationToken: ct).ConfigureAwait(false);
                        if (interval > 0 && i + 1 < count)
                        {
                            await Task.Delay(interval, ct).ConfigureAwait(false);
                        }
                    }
                }
                ,
                _ => throw new NotSupportedException($"unknown handler behavior '{snapshot.Behavior}'"),
            };

            if (snapshot.IsDefault)
            {
                builder.AddCroniqDefaultHandler(handler);
            }
            else if (!string.IsNullOrEmpty(snapshot.Schedule))
            {
                builder.AddCroniqJob(snapshot.JobKey, snapshot.Schedule, handler);
            }
            else
            {
                builder.AddCroniqJob(snapshot.JobKey, handler);
            }
        }
        return builder;
    }

    private static LogLevel ParseLevel(string? level) => (level ?? "info").ToLowerInvariant() switch
    {
        "trace" => LogLevel.Trace,
        "debug" => LogLevel.Debug,
        "info" => LogLevel.Information,
        "warn" => LogLevel.Warning,
        "error" => LogLevel.Error,
        _ => LogLevel.Information,
    };
}

File: sdks/python/tests/conformance/handler_sentinels.py
"""Translate a YAML handler ``behavior`` into a Croniq handler coroutine.

Every conformance binding must implement the same sentinels:
``noop``, ``throw``, ``sleep``, ``log``, ``stream_logs``.
"""

from __future__ import annotations

import asyncio
from collections.abc import Callable, Coroutine
from typing import Any

from croniq_runner import ExecutionContext, HandlerError, LogLevel, Runner
from tests.conformance.case_spec import HandlerSpec


def apply_to(runner: Runner, handlers: list[HandlerSpec]) -> None:
    for spec in handlers:
        handler = _make_handler(spec)
        if spec.is_default:
            runner.set_default_handler(handler)
        else:
            runner.add_handler(spec.job_key, handler, schedule=spec.schedule)


def _make_handler(spec: HandlerSpec) -> Callable[[ExecutionContext], Coroutine[Any, Any, None]]:
    behavior = spec.behavior
    if behavior == "noop":
        async def noop(ctx: ExecutionContext) -> None:
            return None
        return noop

    if behavior == "throw":
        msg = spec.error_message or "thrown by conformance handler"
        async def throw(ctx: ExecutionContext) -> None:
            raise HandlerError(msg)
        return throw

    if behavior == "sleep":
        duration = (spec.duration_ms or 0) / 1000.0
        async def sleep(ctx: ExecutionContext) -> None:
            # Cooperative sleep — fires immediately on cancellation event.
            try:
                await asyncio.wait_for(ctx.cancellation.wait(), timeout=duration)
            except TimeoutError:
                return
            # If we got here, cancellation fired.
            raise asyncio.CancelledError("cancelled during sleep")
        return sleep

    if behavior == "log":
        level = _parse_level(spec.level)
        count = spec.count or 1
        message = spec.message or ""
        async def log(ctx: ExecutionContext) -> None:
            for _ in range(count):
                ctx.logger.log(level, message)
        return log

    if behavior == "stream_logs":
        count = spec.count or 1
        interval = (spec.interval_ms or 0) / 1000.0
        log_level = _parse_log_level(spec.level)
        async def stream(ctx: ExecutionContext) -> None:
            writer = ctx.log_writer
            for i in range(count):
                await writer.write(f"line {i + 1}", level=log_level)
                if interval > 0 and i + 1 < count:
                    await asyncio.sleep(interval)
        return stream

    raise NotImplementedError(f"unknown handler behavior '{behavior}'")


def _parse_level(level: str | None) -> int:
    """Map a YAML level string to a stdlib logging int."""
    import logging
    return {
        "trace": logging.DEBUG,
        "debug": logging.DEBUG,
        "info": logging.INFO,
        "warn": logging.WARNING,
        "error": logging.ERROR,
    }.get((level or "info").lower(), logging.INFO)


def _parse_log_level(level: str | None) -> LogLevel:
    return {
        "trace": LogLevel.TRACE,
        "debug": LogLevel.DEBUG,
        "info": LogLevel.INFO,
        "warn": LogLevel.WARN,
        "error": LogLevel.ERROR,
    }.get((level or "info").lower(), LogLevel.INFO)
